pkg/v1/cmd.go

```diff
@@ -0,0 +1,620 @@
+			"git", "rev-parse", info.Version+"^{}", // get the commit the tag points to, if the tag is its own object
+		), directories[name]))
+		if err != nil {
+			// it's possible that the version is synthetic v0.0.0-date-sha, so check for that
+			var commitFromVersion string
+			commitMatches := syntheticVersionRegex.FindStringSubmatch(commitSha)
+			if len(commitMatches) > 0 {
+				if len(commitMatches[0]) > 1 {
+					commitFromVersion = commitMatches[1]
+				}
+			}
+			if commitFromVersion == "" {
+				return nil, fmt.Errorf("could not determine commit for %s from go.mod version: %w", name, err)
+			}
+			commitSha = commitFromVersion
+		}
+		commitSha = strings.TrimSpace(commitSha)
+		logger.WithFields(logrus.Fields{"repo": name, "commit": commitSha}).Info("resolved latest commit")
+		commit, err := internal.Info(ctx, logger, commitSha, directories[name])
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine commit info: %w", err)
+		}
+		if isUpToDate(ctx, logger, name, directories[name], commit.Hash) {
+			continue
+		}
+		additional, err := detectCarryCommits(ctx, logger, name, directories[name], commit.Hash, mode)
+		if err != nil {
+			return nil, err
+		}
+		target[name] = Config{
+			Target:     commit,
+			Additional: additional,
+		}
+	}
+	return target, nil
+}
+
+func detectNewOperatorControllerCommits(ctx context.Context, logger *logrus.Entry, dir string, mode fetchMode) (*Config, error) {
+	commitSha, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+		"git", "rev-parse", "FETCH_HEAD",
+	), dir))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse upstream HEAD: %w", err)
+	}
+	commitSha = strings.TrimSpace(commitSha)
+	logger.WithFields(logrus.Fields{"repo": "operator-controller", "commit": commitSha}).Info("resolved latest commit")
+	commit, err := internal.Info(ctx, logger, commitSha, dir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine commit info: %w", err)
+	}
+	if isUpToDate(ctx, logger, "operator-controller", dir, commit.Hash) {
+		return nil, nil
+	}
+	additional, err := detectCarryCommits(ctx, logger, "operator-controller", dir, commit.Hash, mode)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve additional commits: %w", err)
+	}
+	return &Config{
+		Target:     commit,
+		Additional: additional,
+	}, nil
+}
+
+func isUpToDate(ctx context.Context, logger *logrus.Entry, repo, dir, commit string) bool {
+	logger = logger.WithField("repo", repo)
+	if _, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+		"git", "merge-base", "--is-ancestor", commit, "main",
+	), dir)); err == nil {
+		logger.WithField("commit", commit).Info("branch already contains target commit, nothing to do")
+		return true
+	}
+	return false
+}
+
+var upstreamCommitRegex = regexp.MustCompile(`^UPSTREAM: (revert: )?(([\w.-]+/[\w-.-]+)?: )?(\d+:|<carry>:|<drop>:)`)
```

Comment:
Would be nice to have unit tests for this

Reply:
I copy pasted this from the openshift-eng repo actually, wanted to ask you about this - should we move that bit from a `main` package so I can import it?

Reply:
Also dove-tailing with that - do we have an exhaustive list of things that it supports? What is `revert: `? Do we cherry-pick those?

Reply:
Importing would be nice. Is it in build-machinery-go/commitchecker?

`revert: ` theoretically tracks reverting a previous `UPSTREAM` commit. Seems like a useful thing to have. If we can ID the thing being reverted, presumably together they cancel each other out?

Reply:
It is not importable since it's in a `package main` in `build-machinery-go` - happy to refactor that if we care.